osu.Game/Utils/BinarySearchUtils.cs

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+
+namespace osu.Game.Utils
+{
+    public class BinarySearchUtils
+    {
+        /// <summary>
+        /// Finds the index of the item in the sorted list which has its property equal to the search term.
+        /// If no exact match is found, the complement of the index of the first item greater than the search term will be returned.
+        /// </summary>
+        /// <typeparam name="T">The type of the items in the list to search.</typeparam>
+        /// <typeparam name="T2">The type of the property to perform the search on.</typeparam>
+        /// <param name="list">The list of items to search.</param>
+        /// <param name="searchTerm">The query to find.</param>
+        /// <param name="termFunc">Function that maps an item in the list to its index property.</param>
+        /// <param name="equalitySelection">Determines which index to return if there are multiple exact matches.</param>
+        /// <returns>The index of the found item. Will return the complement of the index of the first item greater than the search query if no exact match is found.</returns>
+        public static int BinarySearch<T, T2>(IReadOnlyList<T> list, T2 searchTerm, Func<T, T2> termFunc, EqualitySelection equalitySelection = EqualitySelection.FirstFound)
```

Comment:
Not a huge fan of the decision to make this have two typeparams. Would prefer one, and doing the necessary data projection (`termFunc`) in usages before calling this. It just complicates this method for no reason as far as I'm concerned.

I'd understand if the projection was _expensive_, then there'd be an argument that it's reducing the number of invocations of `termFunc`, but no, it's just a `.Time` access. If anything you may get better results by dumping the times off to a list before passing them to this method (*waves hands about cache locality*).

Either that or just move the method back inline where it used to live. I don't really get why generalise this if it's gonna have one caller ever.

Reply:
We probably will and we even already are, but many places just do it locally and that is probably fine. Run a grep for `BinarySearch` on master and you'll see.

After having stepped on similar rakes too many times in the past I want to *see* that something will generalise before allowing someone to generalise. No generalising without at least two usages, preferably more.

Also in general not huge on generics these days either. Generics as implemented in C# introduce _a lot_ of rigidity that is very difficult to root out later, so the least they are used the better in my book.

Reply:
Ok I see we use the built in binary search for lists a lot. I'll move the method back inline.